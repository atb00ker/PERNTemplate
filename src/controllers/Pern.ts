import db from "../models";
import { Request, Response } from 'express';
import * as WebSocket from 'ws';

export const update_websocket = (req: any, res: Response, __: any) => {
  req.ws.clients.forEach(function each(client: any) {
    if (client.readyState === WebSocket.OPEN) {
      client.send(req.params.socket_info);
    }
  })
  return res.status(201).send();
}

export const create_pern = (req: Request, res: Response, __: any) => {
  return db.Pern.create({ name: req.body.name }).then((row: any) => {
    res.status(201).send({ id: row.get('id'), name: row.get('name') });
  });
}

export const get_pern = (_: Request, res: Response, __: any) => {
  return db.Pern.findAll({ attributes: ['id', 'name'] }).then((result: any) => {
    res.status(200).send(result.map((row: any) => row.get()));
  });
}

export const get_pern_details = (req: Request, res: Response, __: any) =>  {
  return db.Pern.findByPk(req.params.pern_id).then((result: any) => {
    if (result)
      res.status(200).send(result.get());
    else
      res.status(200).send();
  });
}

export const put_pern_details = (req: Request, res: Response, __: any)=> {
  db.Pern.findByPk(req.params.pern_id).then((result: any) => {
    if (result) {
      result.update({ name: req.body.name }).then((row: any) => {
        return res.status(200).send(row.get());
      });
    } else {
      db.Pern.create({ id: req.params.pern_id, name: req.body.name }).then((row: any) => {
        return res.status(201).send(row.get());
      });
    }
  });
}

export const delete_pern_details = (req: Request, res: Response, __: any) => {
  return db.Pern.destroy({ where: { id: req.params.pern_id } }).then((deleted: any) => {
    if (deleted)
      res.status(204).send();
    else
      res.status(404).send();
  });
}
