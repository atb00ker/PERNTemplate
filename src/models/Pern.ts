import { Model } from 'sequelize';

interface PernAttributes {
  id: number;
  name: string;
}

module.exports = (sequelize: any, DataTypes: any) => {
  class Pern extends Model<PernAttributes> implements PernAttributes{
    id!: number;
    name!: string;
  }
  Pern.init({
      id: {
        type: DataTypes.UUID,
        defaultValue: DataTypes.UUIDV4,
        allowNull: false,
        primaryKey: true
      },
      name: DataTypes.STRING
    }, {
    sequelize,
    modelName: 'Pern',
  });
  return Pern;
};
