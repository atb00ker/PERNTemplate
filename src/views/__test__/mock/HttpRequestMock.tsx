import axios from 'axios';
import { IResponse } from "../../interfaces/IResponse";
import { IPern } from '../../interfaces/IPern';

export const mockGet_200 = (data: Array<IPern>) => {
  const mockGet = jest.spyOn(axios, 'get');
  mockGet.mockImplementation((_) => {
    const response: IResponse<Array<IPern>> = { data: data, status: 200 }
    return Promise.resolve(response);
  });
}

export const mockGet_400 = () => {
  const mockGet = jest.spyOn(axios, 'get');
  mockGet.mockImplementation((_) => {
    const response: IResponse<Array<IPern>> = { data: [], status: 400 }
    return Promise.resolve(response);
  });
}

export const mockPost_201 = (data: IPern) => {
  const mockGet = jest.spyOn(axios, 'post');
  mockGet.mockImplementation((_, __) => {
    const response: IResponse<IPern> = { data: data, status: 200 }
    return Promise.resolve(response);
  });
}
