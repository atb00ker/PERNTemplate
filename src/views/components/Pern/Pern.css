#submit_btn {
  height: 35px;
}
