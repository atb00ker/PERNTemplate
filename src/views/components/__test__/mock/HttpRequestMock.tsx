import axios from 'axios';
import { IResponse } from "../../../interfaces/IResponse";
import { IPern } from '../../../interfaces/IPern';

export const pern1: IPern = { id: "1", name: "Test#1" };
export const pern2: IPern = { id: "2", name: "Test#2" };

export const mockGet_200 = (data: Array<IPern>) => {
  const mockGet = jest.spyOn(axios, 'get');
  mockGet.mockImplementation((_) => {
    const response: IResponse<Array<IPern>> = { data: data, status: 200 }
    return Promise.resolve(response);
  });
}

export const mockGet_400 = () => {
  const mockGet = jest.spyOn(axios, 'get');
  mockGet.mockImplementation((_) => {
    const response: IResponse<Array<IPern>> = { data: [], status: 400 }
    return Promise.resolve(response);
  });
}

export const mockPost_201 = (data: IPern) => {
  const mockGet = jest.spyOn(axios, 'post');
  mockGet.mockImplementation((_, __) => {
    const response: IResponse<IPern> = { data: data, status: 200 }
    return Promise.resolve(response);
  });
}
