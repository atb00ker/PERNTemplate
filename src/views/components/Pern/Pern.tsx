import React, { useRef, useState, useEffect } from 'react';
import useWebSocket from 'react-use-websocket';
import { postData, getData, deleteData } from "../../services/data";
import { IPern } from "../../interfaces/IPern";
import { ErrorBoundary } from 'react-error-boundary';
import { ErrorFallback } from '../ErrorFallback/ErrorFallback';
import { IResponse } from "../../interfaces/IResponse";
import { Link } from "react-router-dom";
import './Pern.css';

const Pern = () => {
  const pernUrl = "/pern/",
    pernName = useRef(null),
    isInitialMount = useRef(true);
  const [pernInDb, setPernInDb] = useState(Array<IPern>());
  useEffect(() => getPernInDb(), [pernName]);

  const protocol = location.protocol === 'http:' ? 'ws:' : 'wss:';
  const port = location.protocol === 'http:' ? process.env.APP_PORT : '443';
  const { sendMessage, lastMessage, readyState } = useWebSocket(
    `${protocol}//${location.hostname}:${port}`);
  const [messageCollection, setMessageCollection] = useState(Array<string>());

  useEffect(() => {
    // Just send connected message to express server.
    if (!isInitialMount.current) {
      setMessageCollection(messageCollection.concat(lastMessage.data));
      sendMessage('Connected');
    } else
      isInitialMount.current = false;
  }, [lastMessage]);

  const onPernCreateSubmit = (event: React.FormEvent) => {
    event.preventDefault();
    const data = { name: pernName.current.value };
    postData<IResponse<IPern>>(pernUrl, data).then(response => {
      setPernInDb(pernInDb.concat(response.data));
    });
  }

  const getPernInDb = () => {
    getData<IResponse<Array<IPern>>>(pernUrl).then(response => setPernInDb(response.data));
  }

  const deletePern = (uuid: string) => {
    deleteData(`/pern/${uuid}`).then(_ => {
      const updatedPern = pernInDb.filter(record => record.id !== uuid);
      setPernInDb(updatedPern);
    });
  }

  return (
    <React.Fragment>
      <ErrorBoundary FallbackComponent={ErrorFallback}>
        <form onSubmit={(event: React.FormEvent) => onPernCreateSubmit(event)} method="post">
          <input type="text" name="pern_name" id="pern_name" ref={pernName} />
          <button data-testid="pern-submit-btn" id="submit_btn" style={{ margin: '10px' }} type="submit">Submit</button>
        </form>
      </ErrorBoundary>
      <ErrorBoundary FallbackComponent={ErrorFallback}>
        <ul data-testid="pern-list">
          {pernInDb.map(row => (
            <li key={row.id} id={row.id}>
              <span>{row.name}</span>
              <Link style={{ margin: '10px' }} to={`${pernUrl}${row.id}`}>Edit</Link>
              <a href="#" onClick={(event: React.MouseEvent) => { deletePern(`${row.id}`) }}>Delete</a>
            </li>
          ))}
        </ul>
      </ErrorBoundary>
      <ErrorBoundary FallbackComponent={ErrorFallback}>
        {lastMessage?.data && <div>Last Updated: {lastMessage?.data}</div>}
        {lastMessage?.data && messageCollection && <div>Update List: {messageCollection.join(', ')}</div>}
      </ErrorBoundary>
    </React.Fragment>
  );
};

export default Pern;
