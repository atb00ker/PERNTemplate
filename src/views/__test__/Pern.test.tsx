/**
 * @jest-environment jsdom
 */

import '@testing-library/jest-dom';
import 'regenerator-runtime/runtime';
import React from 'react';
import ReactDOM from 'react-dom';
import Pern from '../components/Pern/Pern';
import { render, cleanup, act, waitFor } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { HashRouter as Router } from "react-router-dom";
import { mockGet_200, mockPost_201 } from './mock/HttpRequestMock';
import { pern1, pern2 } from '../../__tests__/mock/PernData';
import * as dotenv from "dotenv";

dotenv.config();
describe('Pern Component', () => {
  beforeEach(() => { mockGet_200([pern1]) });
  afterEach(() => { cleanup });

  it("renders without crashing", () => {
    const root = document.createElement("div");
    ReactDOM.render(<Router><Pern></Pern></Router>, root);
  });

  it("test submit buttom exists and has correct props", async () => {
    let page;
    await act(async () => { page = await render(<Router><Pern></Pern></Router>) });
    const submitBtn = page.getByTestId("pern-submit-btn");
    expect(submitBtn).toBeInTheDocument();
    expect(submitBtn).not.toBeDisabled();
    // expect(submitBtn.getAttribute('style')).toEqual('margin: 10px');
    expect(submitBtn).toHaveStyle('margin: 10px');
    expect(submitBtn).toHaveProperty('id', 'submit_btn');
    expect(submitBtn).toHaveTextContent("Submit");
  });

  it("test submit form and page update", async () => {
    let page;
    mockPost_201(pern2);
    await act(async () => { page = await render(<Router><Pern></Pern></Router>) });
    await waitFor(() => expect(page.getByText('Test#1')).toBeInTheDocument());
    expect(page.getByTestId("pern-list")).not.toHaveTextContent('Test#2');

    await act(async () => {
      const submitBtn = page.getByTestId("pern-submit-btn");
      submitBtn.click();
    });
    await waitFor(() => expect(page.getByText('Test#1')).toBeInTheDocument());
    await waitFor(() => expect(page.getByText('Test#2')).toBeInTheDocument());
  });

  it("snapshot match form", () => {
    const snapTree = renderer.create(<Pern></Pern>);
    expect(snapTree.toJSON()[0]).toMatchSnapshot();
  });
});
