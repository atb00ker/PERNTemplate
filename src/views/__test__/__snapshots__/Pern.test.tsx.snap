// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pern Component snapshot match form 1`] = `
<form
  method="post"
  onSubmit={[Function]}
>
  <input
    id="pern_name"
    name="pern_name"
    type="text"
  />
  <button
    data-testid="pern-submit-btn"
    id="submit_btn"
    style={
      Object {
        "margin": "10px",
      }
    }
    type="submit"
  >
    Submit
  </button>
</form>
`;
